develop/parachain-devs/runtime-development/FRAME/pallet-testing.md

```diff
@@ -0,0 +1,103 @@
+
+#### Test Initialization
+
+Each test starts by initializing the runtime environment, typically using the `new_test_ext()` function, which sets up the mock storage and environment.
+
+```rust
+--8<-- 'code/develop/parachain-devs/runtime-development/FRAME/pallet-testing/test-initialization.rs'
+```
+
+#### Function Call Testing
+
+Call the pallet’s extrinsics or functions to simulate user interaction or internal logic. Use the `assert_ok!` macro to check for successful execution and `assert_err!` to verify that errors are handled properly.
+
+```rust
+--8<-- 'code/develop/parachain-devs/runtime-development/FRAME/pallet-testing/function-call-testing.rs'
+```
+
+#### Storage Testing
+
+After calling a function or extrinsic in your pallet, it's important to verify that the state changes in the pallet's storage match the expected behavior. This ensures that data is updated correctly based on the actions taken.
+
+The following example shows how to test the storage behavior before and after the function call:
+
+```rust
+--8<-- 'code/develop/parachain-devs/runtime-development/FRAME/pallet-testing/storage-testing.rs'
+```
+
+#### Event Testing
+
+It’s also crucial to test the events that your pallet emits during execution. By default, events generated in a pallet using the [`#generate_deposit`](https://paritytech.github.io/polkadot-sdk/master/frame_support/pallet_macros/attr.generate_deposit.html){target=\_blank} macro are stored under the system's event storage key (system/events) as [`EventRecord`](https://paritytech.github.io/polkadot-sdk/master/frame_system/struct.EventRecord.html){target=\_blank} entries. These can be accessed using [`System::events()`](https://paritytech.github.io/polkadot-sdk/master/frame_system/pallet/struct.Pallet.html#method.events){target=\_blank} or verified with specific helper methods provided by the system pallet, such as [`assert_has_event`](https://paritytech.github.io/polkadot-sdk/master/frame_system/pallet/struct.Pallet.html#method.assert_has_event){target=\_blank} and [`assert_last_event`](https://paritytech.github.io/polkadot-sdk/master/frame_system/pallet/struct.Pallet.html#method.assert_last_event){target=\_blank}.
+
+Here’s an example of testing events in a mock runtime:
+
+```rust
+--8<-- 'code/develop/parachain-devs/runtime-development/FRAME/pallet-testing/event-testing.rs'
+```
+
+Some key considerations are:
+
+- Block number - events are not emitted on the genesis block, so you need to set the block number using [`System::set_block_number()`](https://paritytech.github.io/polkadot-sdk/master/frame_system/pallet/struct.Pallet.html#method.set_block_number){target=\_blank} to ensure events are triggered
+- Converting events - use `.into()` when instantiating your pallet’s event to convert it into a generic event type, as required by the system’s event storage
```

Comment:
At the end, two things: 

1. Maybe let's do a "Where to go next"
2. Link to the template pallet's testing setup